### PR TITLE
Add a simple check for spinner recovery time

### PR DESCRIPTION
**Pull request by mzhang28**
Progress towards #12091 

Adds the check for spinner recovery time from the standard [ranking criteria]. This is based on the number of beats after the spinner tail. I imagine it probably won't work very well for spinners that run over red lines, but it's a warning after all and can be ignored in subtle cases

[ranking criteria]: https://osu.ppy.sh/wiki/en/Ranking_criteria/osu%21

image

Not sure if this is the wording or format that we want going forward, feedback is welcome.

If there's any conventions that I'm violating, please let me know since this is my first time working with the codebase. Thanks!

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        public CheckMetadata Metadata { get; } = new CheckMetadata(CheckCategory.Compose, "Spinner recovery time");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateSpinnerRecoveryTooShort(this)
+        };
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            double? recommendedRecoveryTime_ = this.recommendedRecoveryTime(context.InterpretedDifficulty);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:24]: What's with the underscore/non-underscore local names? That's really confusing, please don't use such conventions. Explicit null checks, `.Value` accesses, or even the shut-up operator (`!`) are preferable to this.

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+            // If anything goes, we can skip this check
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:26]: Comment restates code, can be removed (and also accidentally a word?)

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+            yield break;
+        }
+
+        /// <summary>
+        /// Get the recommended number of beats for spinner recovery
+        /// </summary>
+        /// <param name="interpretedDifficulty">The difficulty to interpret the beatmap as</param>
+        /// <returns>Number of beats if there is a guideline for it, null if anything goes</returns>
+        private double? recommendedRecoveryTime(DifficultyRating interpretedDifficulty)
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:58]: Method names should be verb phrases - so `getRecommendedRecoveryTime()`.

**Review by bdach — CHANGES_REQUESTED** [review]
Missing test coverage (see https://github.com/ppy/osu/tree/master/osu.Game.Rulesets.Osu.Tests/Editor/Checks for examples)

**Comment by iptq** [comment]
Woops, working on the tests right now.. didn't see that

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs (hunk 2 above)
Comment (iptq) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:26]: removed

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs (hunk 1 above)
Comment (iptq) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:24]: oh didn't realize .Value was a thing, did that

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs (hunk 3 above)
Comment (iptq) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:58]: gotcha

File: osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            foreach (var (firstObject, secondObject) in context.Beatmap.HitObjects.Zip(context.Beatmap.HitObjects.Skip(1)))
+            {
+                if (firstObject is not Spinner spinner)
+                    continue;
+
+                double timeDifference = secondObject.StartTime - spinner.EndTime;
+
+                var timingPoint = context.Beatmap.ControlPointInfo.TimingPointAt(spinner.EndTime);
+
+                double timeDifferenceInBeats = timeDifference / timingPoint.BeatLength;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:38]: So I figured I'd cross-check this against what is arguably the reference implementation of this, namely the MapsetVerifier implementation of this, and noticed it to be [considerably more involved than this](https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/standard/spread/CheckSpinnerRecovery.cs#L61-L74).

I asked around and apparently this is because the osu! RC says:

> Overall rules and guidelines apply to every kind of osu! difficulty. Rhythm-related rules and guidelines apply to approximately 180 BPM beatmaps with 4/4 time signatures. If your song is drastically faster or slower, some variables might be different, as detailed in [Scaling BPM on the ranking criteria](https://osu.ppy.sh/wiki/en/Ranking_criteria/Scaling_BPM).

And in that link, more specifically:

> For example, osu! Easy difficulties require 4 beats of recovery time after spinners according to the Ranking Criteria. At 180 BPM, that is true, however at 120 BPM, it would be fine to use 3 beats of recovery, and that could gradually shift to 2 beats at 90 BPM.

So it looks like possibly this isn't as simple as you'd think it was from reading the RC alone. While I'm still not entirely understanding why the MapsetVerifier implementation only uses two reference BPM values rather than the whole 90/120/180/240 gamut (maybe @Naxesss can advise here?), we should probably be mirroring MV here as I believe it to be the de-facto standard in these sorts of RC checks.

Comment (Naxesss) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:38]: Ideally, the recovery time would be expressed in milliseconds per difficulty level. That way it is BPM independent. So that's one of the things MV checks against, i.e. 1000 ms, 500 ms, 250 ms for Easy, Normal, and Hard respectively.

However, the RC expresses it in beats to make it easier to check for in the editor. Apparently it also states that
> For the most part, 120 BPM and 240 BPM act as breaking points for when a map's rhythm should be doubled and halved according to guidelines.

[Which results in a BPM-dependent recovery time.](https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/standard/spread/CheckSpinnerRecovery.cs#L161C1-L161C1) Compared to the ideal method, this allows shorter recovery times at high BPMs and requires longer ones at low BPMs (iirc).

So basically, we are bound to get false-positives in one direction and false-negatives in the other depending on who you ask. In this case false-positives are more harmful than false-negatives; we'd like issues to be taken seriously. Therefore MV requires both methods to fail before an issue is generated.

Comment (iptq) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:38]: Oh that's pretty interesting, and I didn't see this check before. I'll port the logic, and update this soon

Comment (iptq) [reply osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs:38]: Oh that's pretty interesting, and I didn't see this check before. I'll port the logic, and update this soon
